### PR TITLE
When the downstream service is not found, Consul provider throws an exception

### PR DESCRIPTION
**Pull request by HonghuiHu**
### Case 1: favicon.ico request
`ResponderMiddleware` generates warning for the route `/favicon.ico`

### Case 2: empty route in service discovery setup by Consul provider
`ExceptionHandlerMiddleware` generates exception for the route `/` in Consul setup.

## Fixes
???

## Proposed Changes
???

### CODE REVIEW

**Comment by HonghuiHu** [comment]
When the downstream service is not found, LeastConnection prompts Warn and RoundRobin throws an exception。
image

File: src/Ocelot/LoadBalancer/LoadBalancers/RoundRobin.cs [hunk 1]
```diff
@@ -27,6 +24,12 @@ public async Task<Response<ServiceHostAndPort>> Lease(HttpContext httpContext)
             var services = await _services();
             lock (_lock)
             {
+                if (services.Count < 1)
+                {
+                    // When the downstream service is not found, LeastConnection prompts Warn and RoundRobin throws an exception, eg: /favicon.ico
```
Comment (raman-m) [inline src/Ocelot/LoadBalancer/LoadBalancers/RoundRobin.cs:29]: Truly speaking this case of icon should not be processed by this method.
Can we filter such cases in upper contexts?

File: src/Ocelot/LoadBalancer/LoadBalancers/RoundRobin.cs [hunk 2]
```diff
@@ -27,6 +24,12 @@ public async Task<Response<ServiceHostAndPort>> Lease(HttpContext httpContext)
...
+                    return new ErrorResponse<ServiceHostAndPort>(new ServicesAreEmptyError($"Services were empty for {httpContext.Request.Path}"));
```
Comment (raman-m) [inline src/Ocelot/LoadBalancer/LoadBalancers/RoundRobin.cs:30]: Does this error returning fix the bug with exception generation by global exception handler?
How will `ErrorResponse` prevent exception generating in upper contexts?

**Review by raman-m — CHANGES_REQUESTED** [review]
Hi Honghui!
Thanks for your interest in Ocelot!

Let's start the discussion...
Currently I am not sure the PR contains right solution for extra requests coming from browsers. 😄

**Review by raman-m — CHANGES_REQUESTED** [review]
Missing unit tests and at least one acceptance test for the user's scenario!

**Comment by raman-m** [comment]
@ggnaegi 
Look at this PR please!
What do you think?..

**Review by ggnaegi — COMMENTED** [review]
@raman-m I think the problem was that round robin behaved differently than other load balancers.
We would expect ``` new ErrorResponse<ServiceHostAndPort>(new ServicesAreEmptyError("There were no services in LoadBalancer")); ``` if no services are available. It seems to me that this is now fixed in the develop branch and this PR could be closed.

``` c#
 var services = await _servicesDelegate?.Invoke() ?? new List<Service>();

            if (services?.Count != 0)
            {
                lock (_lock)
                {
                    if (_last >= services.Count)
                    {
                        _last = 0;
                    }

                    var next = services[_last++];
                    return new OkResponse<ServiceHostAndPort>(next.HostAndPort);
                }
            }

            return new ErrorResponse<ServiceHostAndPort>(new ServicesAreEmptyError($"There were no services in {nameof(RoundRobin)} during {nameof(Lease)} operation."));
```
https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/LoadBalancer/LoadBalancers/RoundRobin.cs

**Comment by raman-m** [comment]
@HonghuiHu
This issue is outdated, and it is already fixed in **develop** branch.
Please, sync your forked repo, and/or download Ocelot v[20.0](../releases) NuGet package

**Pull request closed without merging**